apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -216,49 +229,96 @@ Keep in mind you are not just generating an example; you are generating an actua
       user: openAiUserFromAuthn(authnUserId),
     });
 
-    extractFromCompletion(completion, job);
-    const html = job?.data?.html;
-
-    job.data['initialGenerationErrors'] = [];
-    job.data['prompt'] = userPrompt;
-    job.data['generation'] = completion.choices[0].message.content;
-    job.data['context'] = context;
-    job.data['completion'] = completion;
+    const results = extractFromCompletion(completion, job);
+    const html = results?.html;
 
+    let errors: string[] = [];
     if (html && typeof html === 'string') {
-      const errors = validateHTML(html, false, !!job?.data?.python);
-      job.data['initialGenerationErrors'] = errors;
-      job.data['finalGenerationErrors'] = errors;
-      if (errors.length > 0) {
-        await regenInternal(
-          job,
-          client,
-          authnUserId,
-          userPrompt,
-          `Please fix the following issues: \n${errors.join('\n')}`,
-          html,
-          typeof job?.data?.python === 'string' ? job?.data?.python : undefined,
-          0,
-          false,
-        );
-      }
+      errors = validateHTML(html, false, !!results?.python);
     } else {
+      errors = ['Please generate a question.html file.'];
+    }
+
+    if (userId !== undefined && hasCoursePermissionEdit !== undefined) {
+      const files = {};
+
+      if (results?.html) {
+        files['question.html'] = results?.html;
+      }
+
+      if (results?.python) {
+        files['server.py'] = results?.python;
+      }
+
+      const draftNumber = await queryRow(
+        sql.update_draft_number,
+        { course_id: courseId },
+        z.number(),
+      );
+
+      const client = getCourseFilesClient();
+
+      const qid = `__drafts__/draft_${draftNumber}`;
+
+      await client.createQuestion.mutate({
+        course_id: courseId,
+        user_id: userId,
+        authn_user_id: authnUserId,
+        has_course_permission_edit: hasCoursePermissionEdit,
+        qid,
+        title: `draft ${draftNumber}`,
+        files,
+      });
+
+      await queryAsync(sql.insert_draft_info, {
+        qid,
+        course_id: courseId,
+        creator_id: authnUserId,
+      });
+
+      await queryAsync(sql.insert_prompt_info, {
+        qid,
+        course_id: courseId,
+        prompting_uid: authnUserId,
+        prompt_type: 'initial_prompt',
+        user_prompt: userPrompt,
+        context,
+        response: completion.choices[0].message.content,
+        title: 'temporary draft placeholder (todo: fix)',
+        uuid: `draft_${draftNumber}_todo_fix`,
+        html: results?.html,
+        python: results?.python,
+        errors,
+        completion,
+      });
+      job.data['questionQid'] = qid;
+    }
+
+    job.data.html = html;
+    job.data.python = results?.python;
+
+    if (errors.length > 0 && typeof job.data.questionQid === 'string') {
       await regenInternal(
```

Comment:
This is now up to 13 arguments. I would strongly encourage you to switch to an object with named properties like `generateQuestion` does.

Reply:
done in 55219b95d0f343aa96b9cc8726d142b558d2e55b.